File: ts/src/shareKey.ts
import { ml_kem768 } from "@noble/post-quantum/ml-kem";
import { ml_dsa65 } from "@noble/post-quantum/ml-dsa";
import { arrayBufferToBase64, base64ToArrayBuffer, keyHash } from "./utils.ts";
import { isValidUUIDv7 } from "./core.ts";
import { isValidMasterKeyPrivate, signMasterKey } from "./masterKey.ts";
import type { EncryptedData, shareKey, shareSignKey, Sign } from "./type.ts";

/**
 * 共有キー(ShareKey)を生成する
 * @param masterKey マスターキーの秘密鍵
 * @param sessionUUID セッションUUID
 * @returns 生成された共有キーとその署名、またはエラー時はnull
 */
export async function generateShareKey(
  masterKey: string,
  sessionUUID: string,
): Promise<
  {
    publicKey: string;
    privateKey: string;
    sign: string;
  } | null
> {
  if (!isValidMasterKeyPrivate(masterKey)) {
    return null;
  }
  if (!isValidUUIDv7(sessionUUID)) {
    return null;
  }
  try {
    const key = ml_kem768.keygen();
    const publicKeyBinary = arrayBufferToBase64(
      key.publicKey as unknown as ArrayBuffer,
    );
    const privateKeyBinary = arrayBufferToBase64(
      key.secretKey as unknown as ArrayBuffer,
    );
    const timestamp = new Date().getTime();
    const publicKeyObj: shareKey = {
      keyType: "shareKeyPublic",
      key: publicKeyBinary,
      timestamp: timestamp,
      sessionUuid: sessionUUID,
      algorithm: "ML-KEM-768",
    };
    const privateKeyObj: shareKey = {
      keyType: "shareKeyPrivate",
      key: privateKeyBinary,
      timestamp: timestamp,
      sessionUuid: sessionUUID,
      algorithm: "ML-KEM-768",
    };
    const publicKey = JSON.stringify(publicKeyObj);
    const privateKey = JSON.stringify(privateKeyObj);
    const sign = await signMasterKey(
      masterKey,
      publicKey,
      await keyHash(masterKey),
    );
    if (!sign) {
      return null;
    }
    return {
      publicKey,
      privateKey,
      sign,
    };
  } catch (error) {
    console.error("共有キー生成中にエラー:", error);
    return null;
  }
}

/**
 * 共有キーの公開鍵が有効かどうかを検証する
 * @param key 検証する共有キーの公開鍵
 * @returns 検証結果（true: 有効、false: 無効）
 */
export function isValidShareKeyPublic(key: string): boolean {
  try {
    if (key.length !== 1696) {
      console.log("共有キー公開鍵の長さが無効:", key.length);
      return false;
    }
    const { key: keyBinary, keyType, sessionUuid } = JSON.parse(key);
    if (keyType !== "shareKeyPublic") {
      return false;
    }
    if (!isValidUUIDv7(sessionUuid)) {
      return false;
    }
    const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
    if (keyBinaryArray.length !== 1184) {
      console.log("共有キー公開鍵のバイナリ長が無効:", keyBinaryArray.length);
      return false;
    }
    return true;
  } catch (error) {
    console.error("共有キー公開鍵検証中にエラー:", error);
    return false;
  }
}

/**
 * 共有キーの秘密鍵が有効かどうかを検証する
 * @param key 検証する共有キーの秘密鍵
 * @returns 検証結果（true: 有効、false: 無効）
 */
export function isValidShareKeyPrivate(key: string): boolean {
  try {
    if (key.length !== 3317) {
      console.log("共有キー秘密鍵の長さが無効:", key.length);
      return false;
    }
    const { key: keyBinary, keyType, sessionUuid } = JSON.parse(key);
    if (keyType !== "shareKeyPrivate") {
      return false;
    }
    if (!isValidUUIDv7(sessionUuid)) {
      return false;
    }
    const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
    if (keyBinaryArray.length !== 2400) {
      console.log("共有キー秘密鍵のバイナリ長が無効:", keyBinaryArray.length);
      return false;
    }
    return true;
  } catch (error) {
    console.error("共有キー秘密鍵検証中にエラー:", error);
    return false;
  }
}

/**
 * 共有キーでデータを暗号化する
 * @param key 共有キーの公開鍵
 * @param data 暗号化するデータ
 * @returns 暗号化されたデータのJSON文字列、またはエラー時はnull
 */
export async function encryptDataShareKey(
  key: string,
  data: string,
): Promise<string | null> {
  if (!isValidShareKeyPublic(key)) {
    return null;
  }
  try {
    const { key: keyBinary } = JSON.parse(key);
    const dataArray = new TextEncoder().encode(data);
    const ciphertext = ml_kem768.encapsulate(
      new Uint8Array(base64ToArrayBuffer(keyBinary)),
      new Uint8Array(32),
    );
    const ciphertextString = arrayBufferToBase64(
      ciphertext.cipherText as unknown as ArrayBuffer,
    );
    const keyHashString = await keyHash(key);
    const importedKey = await crypto.subtle.importKey(
      "raw",
      new Uint8Array(ciphertext.sharedSecret),
      "AES-GCM",
      true,
      ["encrypt", "decrypt"],
    );
    const iv = crypto.getRandomValues(new Uint8Array(12));
    const encryptedData = await crypto.subtle.encrypt(
      {
        name: "AES-GCM",
        iv: iv,
      },
      importedKey,
      dataArray,
    );
    const ivString = arrayBufferToBase64(iv as unknown as ArrayBuffer);
    const encryptedDataString = arrayBufferToBase64(encryptedData);
    const result: EncryptedData = {
      keyType: "shareKey",
      keyHash: keyHashString,
      encryptedData: encryptedDataString,
      iv: ivString,
      cipherText: ciphertextString,
    };
    return JSON.stringify(result);
  } catch (error) {
    console.error("共有キーでのデータ暗号化中にエラー:", error);
    return null;
  }
}

/**
 * 共有キーで暗号化されたデータを復号する
 * @param key 共有キーの秘密鍵
 * @param data 復号するデータ
 * @returns 復号されたデータ文字列、またはエラー時はnull
 */
export async function decryptDataShareKey(
  key: string,
  data: string,
): Promise<string | null> {
  if (!isValidShareKeyPrivate(key)) {
    return null;
  }
  try {
    const { key: keyBinary } = JSON.parse(key);
    const { encryptedData, iv, cipherText } = JSON.parse(data);
    if (!cipherText) {
      return null;
    }
    const sharedSecret = ml_kem768.decapsulate(
      new Uint8Array(base64ToArrayBuffer(cipherText)),
      new Uint8Array(base64ToArrayBuffer(keyBinary)),
    );
    const importedKey = await crypto.subtle.importKey(
      "raw",
      new Uint8Array(sharedSecret),
      "AES-GCM",
      true,
      ["encrypt", "decrypt"],
    );
    const decryptedData = await crypto.subtle.decrypt(
      {
        name: "AES-GCM",
        iv: new Uint8Array(base64ToArrayBuffer(iv)),
      },
      importedKey,
      new Uint8Array(base64ToArrayBuffer(encryptedData)),
    );
    return new TextDecoder().decode(decryptedData);
  } catch (error) {
    console.error("共有キーでのデータ復号中にエラー:", error);
    return null;
  }
}

/**
 * 暗号化された共有キーデータが有効かどうかを検証する
 * @param data 検証する暗号化データ
 * @returns 検証結果（true: 有効、false: 無効）
 */
export function isValidEncryptedDataShareKey(data: string): boolean {
  try {
    const { keyType, keyHash, iv, cipherText } = JSON.parse(data);
    const sha256 = new Uint8Array(base64ToArrayBuffer(keyHash));
    if (keyType !== "shareKey") {
      return false;
    }
    if (sha256.length !== 32) {
      return false;
    }
    if (new Uint8Array(base64ToArrayBuffer(iv)).length !== 12) {
      return false;
    }
    if (new Uint8Array(base64ToArrayBuffer(cipherText)).length !== 1088) {
      return false;
    }
    return true;
  } catch (error) {
    console.error("暗号化共有キーデータ検証中にエラー:", error);
    return false;
  }
}

/**
 * 共有署名キー(ShareSignKey)を生成する
 * @param masterKey マスターキーのオブジェクト
 * @param sessionUUID セッションUUID
 * @returns 生成された共有署名キーとその署名、またはエラー時はnull
 */
export async function generateShareSignKey(
  masterKey: {
    publicKey: string;
    privateKey: string;
  },
  sessionUUID: string,
): Promise<
  { 
    publicKey: string; 
    privateKey: string; 
    sign: string;
  } | null
> {
  if (!isValidMasterKeyPrivate(masterKey.privateKey)) {
    return null;
  }
  if (!isValidUUIDv7(sessionUUID)) {
    return null;
  }
  try {
    const seed = crypto.getRandomValues(new Uint8Array(32));
    const key = ml_dsa65.keygen(seed);
    const publicKeyBinary = arrayBufferToBase64(
      key.publicKey as unknown as ArrayBuffer,
    );
    const privateKeyBinary = arrayBufferToBase64(
      key.secretKey as unknown as ArrayBuffer,
    );
    const timestamp = new Date().getTime();
    const publicKey: shareSignKey = {
      keyType: "shareSignKeyPublic",
      key: publicKeyBinary,
      timestamp: timestamp,
      sessionUuid: sessionUUID,
      algorithm: "ML-DSA-65",
    };
    const privateKey: shareSignKey = {
      keyType: "shareSignKeyPrivate",
      key: privateKeyBinary,
      timestamp: timestamp,
      sessionUuid: sessionUUID,
      algorithm: "ML-DSA-65",
    };
    const publicKeyString = JSON.stringify(publicKey);
    const privateKeyString = JSON.stringify(privateKey);
    const sign = await signMasterKey(
      masterKey.privateKey,
      publicKeyString,
      await keyHash(masterKey.publicKey),
    );
    if (!sign) {
      return null;
    }
    return {
      publicKey: publicKeyString,
      privateKey: privateKeyString,
      sign,
    };
  } catch (error) {
    console.error("共有署名キー生成中にエラー:", error);
    return null;
  }
}

/**
 * 共有署名キーの公開鍵が有効かどうかを検証する
 * @param key 検証する共有署名キーの公開鍵
 * @returns 検証結果（true: 有効、false: 無効）
 */
export function isValidShareSignKeyPublic(key: string): boolean {
  try {
    if (key.length !== 2724) {
      console.log("共有署名キー公開鍵の長さが無効:", key.length);
      return false;
    }
    const { key: keyBinary, keyType, sessionUuid } = JSON.parse(key);
    if (keyType !== "shareSignKeyPublic") {
      return false;
    }
    if (!isValidUUIDv7(sessionUuid)) {
      return false;
    }
    const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
    if (keyBinaryArray.length !== 1952) {
      console.log("共有署名キー公開鍵のバイナリ長が無効:", keyBinaryArray.length);
      return false;
    }
    return true;
  } catch (error) {
    console.error("共有署名キー公開鍵検証中にエラー:", error);
    return false;
  }
}

/**
 * 共有署名キーの秘密鍵が有効かどうかを検証する
 * @param key 検証する共有署名キーの秘密鍵
 * @returns 検証結果（true: 有効、false: 無効）
 */
export function isValidShareSignKeyPrivate(key: string): boolean {
  try {
    if (key.length !== 5497) {
      console.log("共有署名キー秘密鍵の長さが無効:", key.length);
      return false;
    }
    const { key: keyBinary, keyType, sessionUuid } = JSON.parse(key);
    if (keyType !== "shareSignKeyPrivate") {
      return false;
    }
    if (!isValidUUIDv7(sessionUuid)) {
      return false;
    }
    const keyBinaryArray = new Uint8Array(base64ToArrayBuffer(keyBinary));
    if (keyBinaryArray.length !== 4032) {
      console.log("共有署名キー秘密鍵のバイナリ長が無効:", keyBinaryArray.length);
      return false;
    }
    return true;
  } catch (error) {
    console.error("共有署名キー秘密鍵検証中にエラー:", error);
    return false;
  }
}

/**
 * 共有署名キーでデータに署名する
 * @param key 共有署名キーの秘密鍵
 * @param data 署名するデータ
 * @param pubKeyHash 公開鍵のハッシュ
 * @returns 署名オブジェクトのJSON文字列、またはエラー時はnull
 */
export function signDataShareSignKey(
  key: string,
  data: string,
  pubKeyHash: string,
): string | null {
  if (!isValidShareSignKeyPrivate(key)) {
    return null;
  }
  try {
    const { key: keyBinary } = JSON.parse(key);
    const dataArray = new TextEncoder().encode(data);
    const signature = ml_dsa65.sign(
      new Uint8Array(base64ToArrayBuffer(keyBinary)),
      dataArray,
    );
    const signString = arrayBufferToBase64(signature as unknown as ArrayBuffer);
    const signResult: Sign = {
      signature: signString,
      keyHash: pubKeyHash,
      keyType: "shareSignKey",
    };
    return JSON.stringify(signResult);
  } catch (error) {
    console.error("共有署名キーでの署名作成中にエラー:", error);
    return null;
  }
}

/**
 * 共有署名キーで署名を検証する
 * @param key 共有署名キーの公開鍵
 * @param sign 署名オブジェクトのJSON文字列
 * @param data 署名されたデータ
 * @returns 検証結果（true: 有効、false: 無効）
 */
export function verifyDataShareSignKey(
  key: string,
  sign: string,
  data: string,
): boolean {
  if (!isValidShareSignKeyPublic(key)) {
    return false;
  }
  try {
    const { key: keyBinary } = JSON.parse(key);
    const signData: Sign = JSON.parse(sign);
    if (signData.keyType !== "shareSignKey") {
      return false;
    }
    const dataArray = new TextEncoder().encode(data);
    return ml_dsa65.verify(
      new Uint8Array(base64ToArrayBuffer(keyBinary)),
      dataArray,
      new Uint8Array(base64ToArrayBuffer(signData.signature)),
    );
  } catch (error) {
    console.error("共有署名キーでの署名検証中にエラー:", error);
    return false;
  }
}

/**
 * 共有署名キーの署名が有効かどうかを検証する
 * @param sign 検証する署名
 * @returns 検証結果（true: 有効、false: 無効）
 */
export function isValidSignShareSignKey(sign: string): boolean {
  try {
    const { keyHash, signature, keyType } = JSON.parse(sign);
    if (sign.length !== 4510) {
      console.log("共有署名キー署名の長さが無効:", sign.length);
      return false;
    }
    if (keyType !== "shareSignKey") {
      return false;
    }
    if (keyHash.length !== 44) {
      console.log("共有署名キー署名のハッシュ長が無効:", keyHash.length);
      return false;
    }
    if (signature.length !== 4412) {
      console.log("共有署名キー署名の署名長が無効:", signature.length);
      return false;
    }
    return true;
  } catch (error) {
    console.error("共有署名キー署名検証中にエラー:", error);
    return false;
  }
}

File: rust/src/share_key.rs
use crate::r#type::{ShareKey, ShareSignKey, EncryptedData, Sign};
use crate::keyutils::{generate_kem_key_pair, generate_dsa65_key_pair};
use crate::crypto::{encrypt, decrypt};
use crate::master_key::{is_valid_master_key_private, sign_master_key};
use crate::core::is_valid_uuid_v7;
use crate::utils::key_hash;
use crate::signature::verify_with_mlds65;
use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine as _;
use chrono::Utc;
use serde_json;

/// ShareKey生成
pub fn generate_share_key(master_priv: &str, session_uuid: &str) -> Option<(String,String,String)> {
    if !is_valid_master_key_private(master_priv) || !is_valid_uuid_v7(session_uuid) { return None }
    let (pub_b64, priv_b64) = generate_kem_key_pair().ok()?;
    let ts = Utc::now().timestamp_millis() as u64;
    let pk = ShareKey{ key_type:"shareKeyPublic".into(), key:pub_b64.clone(), algorithm:"ML-KEM-768".into(), timestamp:ts, session_uuid:session_uuid.into() };
    let sk = ShareKey{ key_type:"shareKeyPrivate".into(), key:priv_b64.clone(), algorithm:"ML-KEM-768".into(), timestamp:ts, session_uuid:session_uuid.into() };
    let pkj = serde_json::to_string(&pk).ok()?;
    let skj = serde_json::to_string(&sk).ok()?;
    let mh = key_hash(master_priv);
    let sign = sign_master_key(master_priv, &pkj, &mh)?;
    Some((pkj,skj,sign))
}
pub fn is_valid_share_key_public(json: &str)->bool {
    serde_json::from_str::<ShareKey>(json).map(|k| k.key_type=="shareKeyPublic" && BASE64.decode(&k.key).map_or(false,|b|b.len()==1184)).unwrap_or(false)
}
pub fn is_valid_share_key_private(json: &str)->bool {
    serde_json::from_str::<ShareKey>(json).map(|k| k.key_type=="shareKeyPrivate"&&BASE64.decode(&k.key).map_or(false,|b|b.len()==2400)).unwrap_or(false)
}
pub fn encrypt_data_share_key(pub_json: &str, data: &str) -> Option<String> {
    let sk = serde_json::from_str::<ShareKey>(pub_json).ok()?;
    if sk.key_type != "shareKeyPublic" { return None; }
    let enc = encrypt(data, &sk.key);
    let ed = EncryptedData {
        key_type: "shareKey".into(),
        key_hash: key_hash(pub_json),
        encrypted_data: enc.encrypted_data,
        iv: enc.iv,
        algorithm: Some(enc.algorithm),
        cipher_text: Some(enc.cipher_text),
    };
    serde_json::to_string(&ed).ok()
}

pub fn decrypt_data_share_key(priv_json: &str, json: &str) -> Option<String> {
    let sk = serde_json::from_str::<ShareKey>(priv_json).ok()?;
    if sk.key_type != "shareKeyPrivate" { return None; }
    let ed: EncryptedData = serde_json::from_str(json).ok()?;
    let ciphertext = ed.cipher_text.as_ref()?;
    Some(decrypt(&ed.encrypted_data, ciphertext, &ed.iv, &sk.key))
}

pub fn is_valid_encrypted_data_share_key(json:&str)->bool {
    serde_json::from_str::<EncryptedData>(json).map(|ed|ed.key_type=="shareKey").unwrap_or(false)
}

/// ShareSignKey生成／検証
pub fn generate_share_sign_key(master_priv:&str, session_uuid:&str)->Option<(String,String,String)> {
    if !is_valid_master_key_private(master_priv) || !is_valid_uuid_v7(session_uuid) { return None }
    let (pub_b64, priv_b64) = generate_dsa65_key_pair().ok()?;
    let ts = Utc::now().timestamp_millis() as u64;
    let pk = ShareSignKey{ key_type:"shareSignKeyPublic".into(), key:pub_b64.clone(), algorithm:"ML-DSA-65".into(), timestamp:ts, session_uuid:session_uuid.into() };
    let sk = ShareSignKey{ key_type:"shareSignKeyPrivate".into(), key:priv_b64.clone(), algorithm:"ML-DSA-65".into(), timestamp:ts, session_uuid:session_uuid.into() };
    let pkj=serde_json::to_string(&pk).ok()?;
    let skj=serde_json::to_string(&sk).ok()?;
    let mh = key_hash(master_priv);
    let sign = sign_master_key(master_priv, &pkj, &mh)?;
    Some((pkj,skj,sign))
}
pub fn is_valid_share_sign_key_public(json:&str)->bool {
    serde_json::from_str::<ShareSignKey>(json).map(|k| k.key_type=="shareSignKeyPublic").unwrap_or(false)
}
pub fn is_valid_share_sign_key_private(json:&str)->bool {
    serde_json::from_str::<ShareSignKey>(json).map(|k| k.key_type=="shareSignKeyPrivate").unwrap_or(false)
}
pub fn sign_data_share_sign_key(priv_json:&str, data:&str, key_hash:&str)->Option<String> {
    let sk: ShareSignKey = serde_json::from_str(priv_json).ok()?;
    if sk.key_type != "shareSignKeyPrivate" { return None; }
    crate::signature::create_signature_object_mlds65(&sk.key, data.as_bytes(), key_hash, "shareSignKey").ok()
}
pub fn verify_data_share_sign_key(pub_json: &str, sign_json: &str, data: &str) -> bool {
    let sk = match serde_json::from_str::<ShareSignKey>(pub_json) {
        Ok(v) => v,
        Err(_) => return false,
    };
    let sig = match serde_json::from_str::<Sign>(sign_json) {
        Ok(v) => v,
        Err(_) => return false,
    };
    if sig.key_type != "shareSignKey" { return false; }
    verify_with_mlds65(&sk.key, data.as_bytes(), &sig.signature)
}
pub fn is_valid_sign_share_sign_key(json:&str)->bool {
    serde_json::from_str::<Sign>(json).map(|s|s.key_type=="shareSignKey").unwrap_or(false)
}
